File: 0003 Longest Substring Without Repeating/solution.cpp
#include <bits/stdc++.h>
using namespace std;

 
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> hash;
        int res = 0;
        for (int i = 0, j = 0; j < s.size(); j ++)
        {
            hash[s[j]] ++;
            while (hash[s[j]] > 1)
                hash[s[i++]] --;
            res = max(res, j - i + 1);
        }
        return res;
    }
};

int main()
{
    Solution s;
    assert( (s.lengthOfLongestSubstring("abcabcbb") == 3) );
    assert( (s.lengthOfLongestSubstring("bbbbb") == 1) );
    return 0;
}

File: 0003 Longest Substring Without Repeating/solution.py
#!python3

class Solution:
    def lengthOfLongestSubstring(self, s):
        dic, res, start, = {}, 0, 0
        for i, ch in enumerate(s):
            if ch in dic:
                res = max(res, i-start)
                start = max(start, dic[ch]+1)
            dic[ch] = i
        return max(res, len(s)-start)

if __name__ == "__main__":
    print(Solution().lengthOfLongestSubstring("abcabcbb") == 3)
    print(Solution().lengthOfLongestSubstring("abba") == 2)
    print(Solution().lengthOfLongestSubstring("pwwkew") == 3)